File: Nodejs/K11.ts
import { IExecutable } from "./IExecutable";
import * as readline from "readline";
import { format } from "util";

export class K11 extends IExecutable {
    Question1(rl?: readline.ReadLine): Promise<string> {
        try {
            let ages = new Array<number>();
            let count = 0;
            const maxNumber = 5;

            let q1 = () => {
                return new Promise((resolve, reject) => {
                    rl.question(format("%d人目の年齢を入力して下さい：", count + 1), (value) => {
                        ages[count] = Number.parseInt(value);
                        resolve();
                    });
                });
            }

            return new Promise(async (resolve, reject) => {
                for (var i = 0; i < maxNumber; i++) {
                    count = i;
                    await q1();
                }
                let average = ages.reduce((prev, cur, index, arr) => {
                    return prev + cur;
                }) / ages.length;
                console.log(format("%d人の平均年齢は%dです。", ages.length, average));
                resolve();
            });
        } catch (e) {
            console.error(e);
        }
    }

    Question2(rl?: readline.ReadLine): Promise<string> {

        let strs = new Array<string>();
        let count = 1;

        let question = () => {
            return new Promise((resolve, reject) => {
                rl.question(format("%dつ目の文字列を入力して下さい。＞", count), (value) => {
                    strs.push(value);
                    count++;
                    resolve();
                });
            });
        };

        return new Promise(async (resolve, reject) => {
            for (let i = 0; i < 2; i++) {
                await question();
            }

            let result = strs.join("");

            console.log(format("%sの文字数は%dです。", result, result.length));

            resolve();
        });


    }

    Question3(rl?: readline.ReadLine): Promise<string> {
        let str: string;

        let question = () => {
            return new Promise((resolve, reject) => {
                rl.question("文字を入力して下さい。＞", (value) => {
                    str = value;
                    resolve();
                });
            });
        };

        return new Promise(async (resolve, reject) => {
            await question();
            let applied: boolean = false;
            if (/^[0-9]$/.test(str)) {
                console.log("この文字は数字です。");
                applied = true;
            }
            if (/^[A-Z]$/.test(str)) {
                console.log("この文字は英大文字です。");
                applied = true;
            }
            if (/^[a-z]$/.test(str)) {
                console.log("この文字は英小文字です。");
                applied = true;
            }
            if (/^[A-Fa-f0-9]$/.test(str)) {
                console.log("この文字は16進数として使える文字です。");
                applied = true;
            }
            if (/^(\.|\,|\!|\?|\:|\;|\"|\'|\-|\(|\))*$/g.test(str)) {
                console.log("この文字は区切り文字です。");
                applied = true;
            }
            if (!applied) {
                console.log("この文字はよくわからない文字です。");
                
            }
            resolve();
        });
    }

    Question4(rl?: readline.ReadLine): Promise<string> {
        let alphabets = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        let inputString = "I am a STUDENT. I am good at GAME PROGRAMMING.";

        let count = 0;
        let _count = 0;
        console.log(`文字列：${inputString}\n`);
        console.log("文字頻度");
        for (var i = 0; i < alphabets.length; i++) {
            for (var j = 0; j < inputString.length; j++) {
                if (alphabets[i] === inputString[j].toUpperCase()) {
                    count++;
                }
            }
            process.stdout.write(`${alphabets[i]}: ${count}\t`);
            count = 0;
            _count++;

            if (_count % 4 === 0) {
                console.log();
            }
        }

        return new Promise((resolve, reject) => {
            resolve();
        });
    }
}

File: Cpp/K11.cpp
#include "K11.h"
#include <cctype>
#include <iostream>
#include <string>

using namespace std;

constexpr int MAX_PEOPLE = 3;

void K11::question1() {
    auto age = 0;
    auto average = 0.0f;

    for (int i = 0; i < MAX_PEOPLE; ++i) {
        cout << i + 1 << "人目の年齢を入力して下さい：";
        cin >> age;
        average += static_cast<float>(age);
    }

    cout.setf(ios_base::fixed, ios_base::floatfield);
    cout.precision(1);
    cout << MAX_PEOPLE << "人の平均年齢は" << average / static_cast<float>(MAX_PEOPLE) << "です。";
}

void K11::question2() {
    cout << "1つ目の文字列を入力して下さい。＞";
    string str1;
    cin.clear();
    cin.ignore();
    getline(cin, str1, '\n');


    cout << "2つ目の文字列を入力して下さい。＞";
    string str2;
    cin.clear();
    getline(cin, str2, '\n');


    str1 += str2;
    cout << str1 << "の文字数は" << str1.length() << "です。";
}

void K11::question3() {
    char chr;
    cout << "文字を入力して下さい。＞";
    cin >> chr;
    auto applied = false;

    if (isdigit(chr)) {
        cout << "この文字は数字です。\n";
        applied = true;
    }

    if (isupper(chr)) {
        cout << "この文字は英大文字です。\n";
        applied = true;
    }

    if (islower(chr)) {
        cout << "この文字は英小文字です。\n";
        applied = true;
    }


    if (isxdigit(chr)) {
        cout << "この文字は16進数として使える文字です。\n";
        applied = true;
    }


    if (ispunct(chr)) {
        cout << "この文字は区切り文字です。\n";
        applied = true;
    }

    if (!applied) {
        cout << "この文字はよくわからない文字です。";
    }
}

void K11::question4() {
    string alphabets = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    string inputString = "I am a STUDENT. I am good at GAME PROGRAMMING.";

    cout << "文字列：" << inputString << "\n\n";

    cout << "文字頻度\n";

    auto count = 0;
    auto _count = 0;

    auto i = 0;

    for (char alphabet: alphabets) {
        for (char j: inputString) {
            if (alphabet == std::toupper(j)) {
                ++_count;
            }
        }

        cout << alphabet << ": " << _count << "\t";
        _count = 0;
        ++count;

        if (count % 4 == 0) {
            printf("\n");
        }
    }
}
